amaru/src/consensus/header_validation.rs

```diff
@@ -0,0 +1,179 @@
+        &mut self,
+        stage: &mut Stage,
+    ) -> Result<WorkSchedule<PullEvent>, WorkerError> {
+        let unit = stage.upstream.recv().await.or_panic()?;
+
+        Ok(WorkSchedule::Unit(unit.payload))
+    }
+
+    async fn execute(&mut self, unit: &PullEvent, stage: &mut Stage) -> Result<(), WorkerError> {
+        match unit {
+            PullEvent::RollForward(point, raw_header) => {
+                info!(?point, "validating roll forward");
+                let header = MultiEraHeader::decode(6, None, raw_header)
+                    .map_err(|e| miette!(e))
+                    .or_panic()?;
+
+                match header {
+                    MultiEraHeader::BabbageCompatible(_) => {
+                        let minted_header = header.as_babbage().unwrap();
+                        let epoch = epoch_for_preprod_slot(minted_header.header_body.slot);
+
+                        // TODO: This is awkward, and should probably belong to the LedgerState
+                        // abstraction? The ledger shall keep track of the rolling nonce and
+                        // provide some endpoint for the consensus to access it.
+                        let epoch_nonce = stage
+                            .epoch_to_nonce
+                            .get(&epoch)
+                            .ok_or(miette!("epoch nonce not found"))
+                            .or_panic()?;
+
+                        let active_slots_coeff: FixedDecimal =
+                            FixedDecimal::from(5u64) / FixedDecimal::from(100u64);
+                        let c = (FixedDecimal::from(1u64) - active_slots_coeff).ln();
+                        let ledger = stage.ledger.lock().unwrap();
+                        let block_validator =
+                            BlockValidator::new(minted_header, &*ledger, epoch_nonce, &c);
+                        block_validator.validate().or_panic()?;
+                        info!(?minted_header.header_body.block_number, "validated block");
+                    }
+                    MultiEraHeader::ShelleyCompatible(_) => {
+                        trace!("shelley compatible header, skipping validation");
+                    }
+                    MultiEraHeader::EpochBoundary(_) => {
+                        trace!("epoch boundary header, skipping validation");
+                    }
+                    MultiEraHeader::Byron(_) => {
+                        trace!("byron header, skipping validation");
+                    }
+                }
+
+                let block = {
+                    info!("fetching block...");
+                    let mut peer_session = stage.peer_session.lock().unwrap();
+                    let client = (*peer_session).blockfetch();
+                    let block = client.fetch_single(point.clone()).await.or_restart()?;
+                    info!("fetched block done.");
+                    block
+                };
+
+                // info!(?block, "fetched block");
```

Comment:
delete line? or update line 145 to add info about the block?

Reply:
These can probably all be switched to `trace!()` or just removed. I only really put them in there when I was debugging the multi-threaded dead-locks to figure out where the code was getting hung up.